### ISSUE TITLE
Send to all ws clients

### ISSUE THREAD

**Issue by hyousef-rusty** [snapview/tungstenite-rs#122]
How can I send msg from the server to ALL opened we clients?

**Comment by najamelan** [comment]
You should have an accept loop for incoming connections. You should loop over the connection and send your message out over each one. Everything in tungstenite deals with a single connection. There is no notion of a group of connections. Looking at the other issue you opened, I'd say this is more a question of how to use rocket. You can probably get an answer fastest on user.rust-lang.org, and otherwise on the rocket issue tracker.

**Comment by application-developer-DA** [comment]
I think @najamelan answered the question very well. If you want to send a message to all connected web-socket clients, you have to iterate over them. `tungstenite` does not hold any specific state which stores all connections on a server side. You can refer to [our server example](https://github.com/snapview/tungstenite-rs/blob/master/examples/server.rs) or even check the "chat" implementation (broadcasts all the messages, i.e. does what you need) from our `tokio-tungstenite` repository [here](https://github.com/snapview/tokio-tungstenite/blob/master/examples/server.rs).

**Comment by brandonros** [comment]
what's the best way to maintain a list of all connections? a vec that you push to on every accept()? it's non-trivial because it's typically cross-thread and you need to handle removing it from the vec on client disconnect from server.

**Comment by agalakhov** [comment]
Usually a `HashMap`. And yes, you have to use `Mutex` or `RwLock` to protect it.

**Comment by brandonros** [comment]
what is the left hand side type (the key) of the hypothetical HashMap here if the right hand side (the value) is the socket itself?

**Comment by agalakhov** [comment]
Some unique user identifier, i.e. network address or a nickname, depending on your requirements. It may be also that the socket is the _left_ side while the right one contains some user metadata. If you don't have any, use `HashSet`.

If you use `mio`-based loop, use the `slab` crate.

**Comment by brandonros** [comment]
```
error[E0277]: the trait bound `tungstenite::WebSocket<std::net::TcpStream>: std::cmp::Eq` is not satisfied
  --> src/server.rs:63:31
   |
63 |             websockets.insert(websocket);
   |                               ^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `tungstenite::WebSocket<std::net::TcpStream>`

```

How would one get around that?

**Comment by agalakhov** [comment]
Use 'HashMap' instead or wrap it into custom struct that supports `Eq`.

**Comment by brandonros** [comment]
@agalakhov I've been racking my brain against this for hours and I apologize in advanced that this isn't the right place to ask this but,

can you show an example (and maybe it should be included somewhere) that shows a `tungstenite` server that is accepting client connections where the server is able to:

1. in a thread, in a loop, read_message() from all clients (no problem, no mut involved, actively documented in the examples)

but also

2. in another thread, in a loop, write_message() to a client

I keep running into all kinds of borrowing/reference/moving/copying woes with the WebSocket that I just do not understand.

A good example is:

imagine you are trying to broker say PostgresSQL or Redis pub/sub traffic to websocket clients.

**Comment by agalakhov** [comment]
@brandonros The reason you have difficulties here is, you're inventing a scheduler. This is not easy. I bet you created threads thinking that this will enable you to send and receive at the same time. This is not a solution.

So, short answer: use Tokio. It provides a scheduler.

**Comment by brandonros** [comment]
I was able to avoid bringing in a scheduler by extracting the stream inside of a WebSocket, then try_clone()ing it into 2 hashmaps: 1 for reading, 1 for writing.

**Comment by agalakhov** [comment]
And how many streams do you have now? What does the cloning of the stream mean?

Why don't you just use Tokio? What you're doing looks like reinventing a square wheel.

**Comment by brandonros** [comment]
i'm lead to believe there is only that one underlying stream (the connection from the client to the server), and try_clone() is just a bunch of "references" being passed around

**Comment by avindra** [comment]
I think @brandonros may be trying to do what I'm doing: use a websocket server to pass a message to a single client. Broadcasting is only a means of getting the message to the single client that we care about. "Broadcasting" is making the statement that we don't care "who" is listening.

Tokio's chat example maintains a list of  `peers`, which is a (`HashMap<SocketAddr, Tx>`) as @agalakhov [alluded to earlier](https://github.com/snapview/tungstenite-rs/issues/122#issuecomment-653806431).

https://github.com/tokio-rs/tokio/blob/066965cd59d01fd9d999152e32169a24dfe434fa/examples/chat.rs

That looks like it'll work... but even then, I wish there were a more basic example. For example, if the client **does not** send messages, only one-way communication is needed, which _should_ lead to a small implementation.

**Comment by avindra** [comment]
I've got something working now, and I can see how my struggles were mostly due to a lack of understanding about Rust. I ended up using `Arc` and `Mutex`, and was able to avoid `std::sync::mpsc::channel` by adding the `Send` attribute to the closure I'm returning. 

Here is a module with a function named `create_broadcaster` that returns a closure that sets up a Websocket server. The returned closure accepts a single `String`, which is broadcasted to a single client<sup>†</sup> (the most recently connected one).

<sup>†</sup> You should  modify this to work with more than just the first index if you need  to broadcast to all connected clients.

```rust
use std::{net::TcpListener, thread::spawn};

use std::sync::{Arc, Mutex};
use std::collections::HashMap;

use tungstenite::{ accept, Message };

type PeerMap = Arc<Mutex<HashMap<u8, tungstenite::WebSocket<std::net::TcpStream>>>>;

pub type Broadcaster = dyn Fn(String) + Send;

/**
 * Create and run a websocket server.
 * Return a closure that
 * is capable of broadcasting
 * to its connected clients.
 */
pub fn create_broadcaster() -> Box<Broadcaster> {
	let server = TcpListener::bind("127.0.0.1:8080").unwrap();
	let peers: PeerMap = Arc::new(Mutex::new(HashMap::new()));

	println!("collecting peers");
	// collect peer list on another thread
	let peers1 = peers.clone();
	spawn(move || {
		for stream in server.incoming() {
				let websocket = accept(stream.unwrap()).unwrap();
				println!("Registered another client");
				peers1.lock().unwrap().insert(0, websocket);
		}
	});

	println!("creating closure");
	// return a broadcaster for immediate use
	return Box::new(move |msg| {
		for sock in peers.lock().unwrap().iter_mut() {
			let msg1 = msg.clone();
			let out = Message::from(msg1);
			sock.1.write_message(out);
		}
	});
}
```

Then, in another module, you'll be able to:

```rust
// use crate::path::to::create_broadcaster;

pub fn run() -> () {
  let broadcaster = create_broadcaster();

  // At this point, the websocket server is up and running on a thread somewhere...
  // but we can broadcast messages due to the api
  // we created in our closure

  broadcaster("Hello, world!");
}
```

[This video](https://www.youtube.com/watch?v=2mwwYbBRJSo) by @rylev (and [code](https://github.com/rylev/rust-workshop/blob/master/part-3/solution/src/lib.rs)) helped clarify the use of the concurrent/parallel primitives like `Arc`, `Mutex`, `+ Send` etc.

Now if there are any experts, please tell me if I did something wrong :sweat_smile:

**Comment by agalakhov** [comment]
So basically you lock the whole list of websockets for the entire time of sending. This has several downsides. First, you're unable to receive any messages while sending. Second, you're talking to clients one-by-one and not in parallel. This means that if one of these clients is slow, the rest must wait. If you have, say, 1000 clients and each of them requires 1 second, the `for` would require 1000 seconds! This could be done much faster taking into account that all clients are completely independent, so one could talk to all of them in parallel. This could be done with threads (quite inefficient), Tokio or Mio.

Also please don't forget that, in many cases, you have to send to _most_ and not to _all_. For example, broadcast the received message to all except the sender. Or broadcast to all users who entered password, ignore anonymous users. It is wise to support `multicast` instead of broadcast.

**Comment by avindra** [comment]
I am ok with the peer list being locked, as I am the sole user 🙂 (its for a experiment / fun project).

But thanks for the advice, it looks like I can get some speedups still even for my simple use case.

**Comment by badonyt** [comment]
@avindra  i saw the snippet but i could understand how to implement it?

**Comment by badonyt** [comment]
This is my code but i dont understand how the function could fit in this mess

```
use std::net::TcpListener;
use std::thread::spawn;
use tungstenite::accept;
use std::sync::Mutex;
use std::sync::Arc;
struct User {
    x: i64,
    y: i64,
    id: i64
}

fn main() {
             // Arc lets us share the Mutex, Mutex lets us share the Vec
    let vec = Arc::new(Mutex::new(Vec::new()));
    vec.lock().map(|mut vec| {
        vec.push(User { x: 1, y: 1, id: 32 });
    });

    let server = TcpListener::bind("127.0.0.1:9001").unwrap();
   while true == true{





   //BROADCAST


}
    for stream in server.incoming() {
        let other_vec = vec.clone();
        spawn (move || {
            let vec = other_vec; // give this thread a reference to the Mutex containing the Vec
            let mut websocket = accept(stream.unwrap()).unwrap();
            loop {
                let msg = websocket.read_message().unwrap();
                let s_slice: &str = &msg.to_string();  // okay
                let parsed = json::parse(s_slice).unwrap();
                if parsed["type"] == "initial" {
                    let new_data = &parsed["data"];
                    vec.lock().map(|mut vec| {
                        // you can modify vec in here
                       vec.push(User { x: new_data["x"].as_i64().unwrap(), y: new_data["x"].as_i64().unwrap(), id: new_data["id"].as_i64().unwrap()});
                       println!("{}", vec[vec.len() - 1].id);
                    });
                    
                }else if parsed["type"] == "keys"{
                    let new_data = &parsed["data"];
                    println!("{}", new_data["what"]);
                    vec.lock().map(|mut vec| {
                        let mut y = 0;
                        while y < vec.len() {
                            if vec[y].id == new_data["id"]{
                                println!("vec is {} and new is {}", vec[y].id, new_data["id"]);
                                if new_data["what"] == "d"{
                                    if vec[y].x != 400{
                                        vec[y].x = vec[y].x + 5
                                    }
                                }else if new_data["what"] == "a"{
                                    if vec[y].x != 0{
                                        vec[y].x = vec[y].x - 5
                                    }
                                }
                            }
                            println!("the current number is {} and {}", y, vec[y].x );
                            y=y+1;
                        }
                    });
                     
                }
                // We do not want to send back ping/pong messages.
                if msg.is_binary() || msg.is_text() {
                    //vec.lock().map(|mut vec| {
                        // you can modify vec in here
                      //  websocket.write_message(tungstenite::Message::Text(vec[0].id.to_string())).unwrap();
                    //});
                }
            }
        });
    }

}
```

**Comment by badonyt** [comment]
(I dont want you to outright give me the answer, just tell me what i could do to make it work)

**Comment by najamelan** [comment]
I haven't looked to much in detail at the code presented here, so forgive me if I'm not considering all constraints. In general, async code has a lot of options for doing things concurrently. As such, the simplest concept when you want to deal with multiple websockets is to have each receiver in a separate task that you spawn. There is structures that can manage the collection for you and poll it to completion, like `FuturesUnordered` from the futures library, or the slightly more advanced `async_nursery` (disclaimer, I wrote that). 

If you want to manage your incoming message centrally, you can have each of these tasks send incoming messages into an mpsc channel. 

For sending, in general it will be easiest to hold all senders in one place. That way you don't have issues with locking et al. Again, if you want to be able to send from all over your code base, just do yourself a favor use an mpsc channel so you can hand out plenty of senders and the channel will deal with being a lockless queue so you don't have to.

**Comment by badonyt** [comment]
I dont understand how a [mpsc](https://doc.rust-lang.org/std/sync/mpsc/) will help me?

**Comment by najamelan** [comment]
> I dont understand how a mpsc will help me?

For the sending or receiving side? I would use an async mpsc, like the one from the futures library. It will basically take care of the locking for you. Eg. You will need a `&mut` to be able to send on the WS, but if you want to be able send from several places in your code, you can just leave on async task that holds your vec of WS senders, and you can clone the channel sender as much as you want and hand it to any code. You don't need any locks.

For the receiving end, it allows you to have each receiver in a separate task but process your messages in one place if that's more convenient. Eg, you could wrap them in a struct with another field (eg. peer ip) to differentiate which connection a message comes from.

You don't have to use a channel, but it's often convenient. It will depend on your exact requirements what you'll use.

**Comment by badonyt** [comment]
hey i undestood the code in the snippet, but does it have something to when the user disconects from the websocket connection

**Comment by badonyt** [comment]
Hey i was trying a test server,
```rust
use std::sync::{Arc, Mutex};
use std::collections::HashMap;

use tungstenite::{ accept, Message };

type PeerMap = Arc<Mutex<HashMap<u8, tungstenite::WebSocket<std::net::TcpStream>>>>;


use std::{net::TcpListener, thread::spawn};

use tungstenite::{
    accept_hdr,
    handshake::server::{Request, Response},
};

fn main() {
    env_logger::init();
    let server = TcpListener::bind("127.0.0.1:3012").unwrap();
    let peers: PeerMap = Arc::new(Mutex::new(HashMap::new()));
    let peers1 = peers.clone();
    for stream in server.incoming() {
        spawn(move || {
            for stream in server.incoming() {
				let websocket = accept(stream.unwrap()).unwrap();
				println!("Registered another client");
				peers1.lock().unwrap().insert(0, websocket);
		}
            
            let callback = |req: &Request, mut response: Response| {
                println!("Received a new ws handshake");
                println!("The request's path is: {}", req.uri().path());
                println!("The request's headers are:");
                for (ref header, _value) in req.headers() {
                    println!("* {}", header);
                }

                // Let's add an additional header to our response to the client.
                let headers = response.headers_mut();
                headers.append("MyCustomHeader", ":)".parse().unwrap());
                headers.append("SOME_TUNGSTENITE_HEADER", "header_value".parse().unwrap());

                Ok(response)
            };
            let mut websocket = accept_hdr(stream.unwrap(), callback).unwrap();

            loop {
                let msg = websocket.read_message().unwrap();
                if msg.is_binary() || msg.is_text() {
                    websocket.write_message(msg).unwrap();
                    for sock in peers.lock().unwrap().iter_mut() {
                        let msg1 = msg.clone();
                        let out = Message::from(msg1);
                        sock.1.write_message(out);
                    }
                }
            }
        });
    }
}
```

which doesnt work, because of this 
![image](https://user-images.githubusercontent.com/78484256/213016047-22c6b543-5fe7-4190-b71a-503626c7b0e1.png)

**Comment by badonyt** [comment]
do you know what to do i do, with the other errors?

**Comment by badonyt** [comment]
(gotta go to sleep, ill check your message tomorrow)

**Comment by najamelan** [comment]
```rust
    for stream in server.incoming() {
        spawn(move || {
            for stream in server.incoming() {
```
You shouldn't loop over the incoming streams twice, let alone nested...

In general you need to learn rust basics. Ownership, references, moves etc. Read [the book](https://doc.rust-lang.org/book/).
